Component parameter assignment guidance (#21555)
dotnet/AspNetCore.Docs: aspnetcore/blazor/components/index.md

  ```razor
  <ChildComponent Title="@GetTitle()">
      Title from method.
  </ChildComponent>
  
  @code {
      private string GetTitle()
      {
          return "Panel Title from Parent";
      }
  }
  ```
  
* To assign the result of an [implicit C# expression](xref:mvc/views/razor#implicit-razor-expressions) in the parent component to a child component's parameter, prefix the implicit expression with an `@` symbol. To assign the current local date in long format with <xref:System.DateTime.ToLongDateString%2A> in the parent component, assign `@DateTime.Now.ToLongDateString()` to the child's parameter. To assign a `person` object's `Name` property, assign `@person.Name`. The following example demonstrates these concepts:

  ```razor
  <ChildComponent Title="@DateTime.Now.ToLongDateString()">
      Title from implicit Razor expression.
  </ChildComponent>
  
  <ChildComponent Title="@person.Name">
      Title from implicit Razor expression.
  </ChildComponent>

>   ```

Reviewer: The above is all absolutely fine and good now, thanks! I do still wonder why we're splitting it into these three cases, since there's nothing in the product code that distinguishes them. As far as the Razor language is concerned, all three of these are simply "using a C# expression", and we don't care whether the expression is a field/property/method-call/complex-expression.

You could even combine all three into a single example, similar to:

----

To assign the result of a C# expression to a child component parameter, use the `@` symbol. As you see in the following example, the expression can be a field, property, method call, or other complex evaluation.

```razor
<ChildComponent Visible="@IsVisible" Title="@ComputeTitle()"  ExpiryDate="@DateTime.Now.AddDays(7)" />
```

Author: I'm only concerned on the **_gotcha_** 😈 scenarios that work in a Razor page but not a Razor component parameter assignment. IMO, we *might* want to leave those two split out and combine the rest.

I'll work it again Tuesday morning and ping u back.

Author: @SteveSandersonMS ... I'm not so sure about making this change. Here's the deal ... readers are cutting and pasting our code from these topics directly into their own test app to play with the scenarios. Your example changes to combine (and including your prompt to not `ToString` the DT) break that educational experience because they can't use the `ChildComponent` component any longer to perform the testing/hacking/playing with our code. I don't think we should necessarily expand (and complicate) that child component. I'll work with it a bit more and see what I can come up with. I'll at least try combining the simple use cases into a single bullet and example.

>   Unlike in Razor pages (`.cshtml`), Blazor can't perform asynchronous work in a Razor expression while rendering a component. Instead, asynchronous work is performed during one of the [asynchronous component lifecycle events](xref:blazor/components/lifecycle). The following Razor syntax is **not** supported:

Reviewer: Just to clarify that it's for a good reason, not just a technological limitation:

```suggestion
  Unlike in Razor pages (`.cshtml`), Blazor can't perform asynchronous work in a Razor expression while rendering a component. This is because Blazor is designed for rendering interactive UIs, and in an interactive UI, there must always be something on the screen, so it doesn't make sense to block the rendering flow. Instead, asynchronous work is performed during one of the [asynchronous component lifecycle events](xref:blazor/components/lifecycle), after which the UI may be updated. The following Razor syntax is **not** supported:
```

  
  ```razor
  <ChildComponent Title="@await ...">
      ...
  </ChildComponent>
  ```
  
  The code in the preceding example generates an IntelliSense error in Visual Studio and a *compiler error* if the app is built:
  
  > The 'await' operator can only be used within an async method. Consider marking this method with the 'async' modifier and changing its return type to 'Task'.

  To obtain a value for the `Title` parameter in the preceding example asychronously, the component can use the [`OnInitializedAsync` lifecycle event](xref:blazor/components/lifecycle#component-initialization-methods), as the following example demonstrates:
  
  ```razor
  <ChildComponent Title="@title">
      Title from implicit Razor expression.
  </ChildComponent>
  
  @code {
      private string title;
      
      protected override async Task OnInitializedAsync()
      {

>           title = await Task.FromResult<string>("Panel Title from Parent");

Reviewer: The `Task.FromResult` snippet may be confusing because people might think they are supposed to use that in their code for some reason. It might be easier to follow like this:

```suggestion
          title = await ...;
```

... like you have in the Razor snippet above.

      }
  }
  ```
  
* To assign the result of an [explicit C# expression](xref:mvc/views/razor#explicit-razor-expressions) in the parent component to a child component's parameter, surround the expression in parentheses with an `@` symbol prefix. To calculate a date that's one week in the past for assignment to a child's parameter, use the syntax `Title="@((DateTime.Now - TimeSpan.FromDays(7)).ToString())"`, as seen in the following parent component:

  ```razor

>   <ChildComponent Title="@((DateTime.Now - TimeSpan.FromDays(7)).ToString())">

Reviewer: Putting `.ToString()` on this is a bit weird. Suggestion:

```suggestion
  <ChildComponent ShowItemsSinceDate="@(DateTime.Now - TimeSpan.FromDays(7))">
```

Author: btw - This happened because my parameter in the child for this example is marked `string` because I was trying to pass a **_title string_**. I probably should've made it clearer in the text what I was trying to do. It's not really a mistake given what I was doing; but given that you're changing the example here to pass a DT, it's not going to be a factor.

  </ChildComponent>
  ```

  Use of an explicit expression to concatenate text with an expression result for assignment to a parameter is **not** supported. The following example seeks to concatenate the text "SKU-" with a product stock number (`SKU` property, "Stock Keeping Unit") provided by a parent component's `product` object. Although this syntax is supported in a Razor page (`.cshtml`), it isn't valid for assignment to the child's `Title` parameter.
  
  ```razor
  <ChildComponent Title="SKU-@(product.SKU)">
      Title from composed Razor expression. This doesn't compile.
  </ChildComponent>
  ```
  
  The code in the preceding example generates an IntelliSense error in Visual Studio and a *compiler error* if the app is built:
  
  > Component attributes do not support complex content (mixed C# and markup).
  
  To support the assignment of a composed value, use a method, field, or property. The following example performs the concatination of "SKU-" and a product's stock number in the C# method `GetTitle`:
  
  ```razor
  <ChildComponent Title="@GetTitle()">
      Composed title from method.
  </ChildComponent>
  
  @code {
      private Product product = new Product();

>       private string GetTitle()
>       {
>           return $"SKU-{product.SKU}";
>       }

Reviewer: Not a big deal, but might be nice:

```suggestion
      private string GetTitle() => $"SKU-{product.SKU}";
```

      
      private class Product
      {
          public string SKU { get; set; } = "12345";
      }
  }
  ```
  
For more information, see [Razor syntax reference for ASP.NET Core](xref:mvc/views/razor).

> [!WARNING]
> Don't create components that write to their own *component parameters*, use a private field instead. For more information, see the [Overwritten parameters](#overwritten-parameters) section.

#### Component parameters should be auto-properties

Component parameters should be declared as *auto-properties*, meaning that they shouldn't contain custom logic in their getters or setters. For example, the following `StartData` property is an auto-property:

```csharp
[Parameter]
public DateTime StartData { get; set; }
```

> Don't place custom logic in the `get` or `set` accessor because component parameters are purely intended for use as a channel for a parent component to flow information to a child component. If a setter of a child component property contains logic that causes rerendering of the parent component, an *infinite rendering loop results*.

Reviewer: I don't know why *infinite rendering loop results* is in italics here, but will leave that to you :)

Author: Just to emphasize that it will :boom:. I'll remove it.

```suggestion
Don't place custom logic in the `get` or `set` accessor because component parameters are purely intended for use as a channel for a parent component to flow information to a child component. If a setter of a child component property contains logic that causes rerendering of the parent component, an infinite rendering loop results.
```

If you need to transform a received parameter value:

> * Leave the parameter property as a pure auto-property to represent the initial raw data.

Reviewer: ```suggestion
* Leave the parameter property as a pure auto-property to represent the supplied raw data.
```

  The following parent component displays four instances of the preceding `ChildComponent` component and sets their `Title` parameter values to:

  * The value of the `title` field.
  * The result of the `GetTitle` C# method.
  * The current local date in long format with <xref:System.DateTime.ToLongDateString%2A>.
  * The `person` object's `Name` property.

  `Pages/ParentComponent.razor`:
  
  ```razor
  <ChildComponent Title="@title">
      Title from field.
  </ChildComponent>
  
  <ChildComponent Title="@GetTitle()">
      Title from method.
  </ChildComponent>
  
  <ChildComponent Title="@DateTime.Now.ToLongDateString()">
      Title from implicit Razor expression.
  </ChildComponent>
  
  <ChildComponent Title="@person.Name">
      Title from implicit Razor expression.
  </ChildComponent>
  
  @code {
      private string title = "Panel Title from Parent";
      private Person person = new Person();
      
      private string GetTitle()
      {
          return "Panel Title from Parent";
      }
      
      private class Person
      {
          public string Name { get; set; } = "Dr. Who";
      }
  }
  ```
  

>   Unlike in Razor pages (`.cshtml`), Blazor can't perform asynchronous work in a Razor expression while rendering a component. This is because Blazor is designed for rendering interactive UIs. In an interactive UI, the screen must always display something, so it doesn't make sense to block the rendering flow. Instead, asynchronous work is performed during one of the [asynchronous lifecycle events](xref:blazor/components/lifecycle). After the component is rendered, the UI may be updated. The following Razor syntax is **not** supported:

Reviewer: ```suggestion
  Unlike in Razor pages (`.cshtml`), Blazor can't perform asynchronous work in a Razor expression while rendering a component. This is because Blazor is designed for rendering interactive UIs. In an interactive UI, the screen must always display something, so it doesn't make sense to block the rendering flow. Instead, asynchronous work is performed during one of the [asynchronous lifecycle events](xref:blazor/components/lifecycle). After each asynchronous lifecycle event, the UI may be updated again. The following Razor syntax is **not** supported:
```

  
  ```razor
  <ChildComponent Title="@await ...">
      ...
  </ChildComponent>
  ```
  

>   The code in the preceding example generates an IntelliSense error in Visual Studio and a *compiler error* if the app is built:

Reviewer: We support non-VS scenarios too (e.g., VS Code or command line) so best not to refer to VS or IntelliSense here.

```suggestion
  The code in the preceding example generates a *compiler error* if the app is built:
```

      </div>
  </div>

  @code {
      [Parameter]
      public DateTime ShowItemsSinceDate { get; set; }

      [Parameter]
      public RenderFragment ChildContent { get; set; }
  }
  ```
  
  The following parent component calculates a date with an explicit C# expression that's one week in the past for assignment to the child's `ShowItemsSinceDate` parameter.
  
  `Pages/ParentComponent.razor`:

  ```razor
  <ChildComponent ShowItemsSinceDate="@(DateTime.Now - TimeSpan.FromDays(7))">
      Title from explicit Razor expression.
  </ChildComponent>
  ```

  Use of an explicit expression to concatenate text with an expression result for assignment to a parameter is **not** supported. The following example seeks to concatenate the text "SKU-" with a product stock number (`SKU` property, "Stock Keeping Unit") provided by a parent component's `product` object. Although this syntax is supported in a Razor page (`.cshtml`), it isn't valid for assignment to the child's `Title` parameter.
  
  ```razor
  <ChildComponent Title="SKU-@(product.SKU)">
      Title from composed Razor expression. This doesn't compile.
  </ChildComponent>
  ```
  

>   The code in the preceding example generates an IntelliSense error in Visual Studio and a *compiler error* if the app is built:

Reviewer: ```suggestion
  The code in the preceding example generates a *compiler error* if the app is built:
```

Again, it's not important that VS gives you a preview of the compiler errors that you're going to get if you build. That's just a VS feature.
